exercises/dominik_arominski/ex02_oo_basics/src/main.cpp

```diff
@@ -0,0 +1,154 @@
+#include <iostream>
+#include <string>
+#include <cmath>
+#include <vector>
+#include <algorithm>
+
+class Shape {
+
+protected:
+  std::string type;
+  int nSides;
+
+public:
+  
+  std::string GetType() const { return type;};
+  int GetSides() const { return nSides;};
+  
+  virtual double CalculatePerimeter() const { return 0;};  
+  virtual double CalculateArea() const { return 0;};  
+
+  friend std::ostream& operator<<(std::ostream& out, const Shape& shape) {
+    return out << shape.GetType() << "\t" << shape.GetSides() << "\t" << shape.CalculateArea() << "\t" << shape.CalculatePerimeter();
+  }
+  
+};
+
+class Triangle : public virtual Shape {
+private:
+  double height;
+  double base;
+public:
+  Triangle(double fHeight, double fBase) : height(fHeight), base(fBase) {
+    type = "Triangle";
+    nSides = 3;
+  }
+  
+  double CalculatePerimeter() const {
+    return base + 2.0 * sqrt(pow(height, 2) + 0.25 * pow(base, 2));
+  }
+  double CalculateArea() const {
+    return 0.5 * base * height;
+  }
+
+};
+
+class Circle : public virtual Shape {
+private:
+  double radius;
+public:
+  Circle(double fRadius=0) : radius(fRadius) {
+    type = "Circle";
+    nSides = 0;
+  }
+
+  double CalculatePerimeter() const {
+    return 2.0 * M_PI * radius;
+  }
+  double CalculateArea() const {
+    return M_PI * pow(radius, 2);
+  }
+};
+
+class Rectangle : public virtual Shape {
+protected:
```

Comment:
Those were private for the other shape, I see that you intent to subclass for the Square, but it would be nice to have all the shapes consistent in this sense.